### PR TITLE
Support fetching observability from ctx.

### PR DESCRIPTION
**Pull request by k4leung4**
Signed-off-by: Kenny Leung <kleung@chainguard.dev>

# Changes

This enables callers of sharedmain to infuse ctx with a metrics.ObservabilityConfig instead of relying on the API server for it.

- 🎁  Add new feature

/kind enhancement

Fixes #

**Release Note**

```release-note
The observability configurations can now be configured with sharedmain.WithObservabilityConfig.

```

**Docs**

```docs

```

### CODE REVIEW

File: injection/sharedmain/main.go [hunk 1]
```diff
@@ -104,6 +104,25 @@ func GetLeaderElectionConfig(ctx context.Context) (*leaderelection.Config, error
...
+// 3. defaults (if not found).
+func GetObservabilityConfig(ctx context.Context) (*metrics.ObservabilityConfig, error) {
+	if cfg := metrics.GetObservabilityConfig(ctx); cfg != nil {
+		return cfg, nil
+	}
+
+	observabilityConfigMap, err := kubeclient.Get(ctx).CoreV1().ConfigMaps(system.Namespace()).Get(ctx, metrics.ConfigMapName(), metav1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		return metrics.NewObservabilityConfigFromConfigMap(nil)
+	} else if err != nil {
```
[the remark this answers is not in the document]
Comment (k4leung4) [reply injection/sharedmain/main.go]: fixed.

File: injection/sharedmain/main_test.go [hunk 2]
```diff
@@ -103,3 +104,19 @@ func TestWithLeaderElectionConfig(t *testing.T) {
...
+	}
+	ctx := metrics.WithConfig(context.Background(), want)
+
+	got, err := GetObservabilityConfig(ctx)
+	if err != nil {
+		t.Fatalf("GetObservabilityConfig() = %v", err)
+	}
+	if diff := cmp.Diff(got, want); diff != "" {
+		t.Errorf("(-got, +want) = %s", diff)
+	}
```
[the remark this answers is not in the document]
Comment (k4leung4) [reply injection/sharedmain/main_test.go:127]: added

File: injection/sharedmain/main.go [hunk 3]
```diff
@@ -104,6 +104,25 @@ func GetLeaderElectionConfig(ctx context.Context) (*leaderelection.Config, error
...
+		return nil, err
+	}
+	return metrics.NewObservabilityConfigFromConfigMap(observabilityConfigMap)
+
```
[the remark this answers is not in the document]
Comment (k4leung4) [reply injection/sharedmain/main.go]: done

File: injection/sharedmain/main.go [hunk 4]
```diff
@@ -249,6 +268,14 @@ func MainWithConfig(ctx context.Context, component string, cfg *rest.Config, cto
 			leaderElectionConfig.GetComponentConfig(component))
 	}
 
+	observabilityConfig, err := GetObservabilityConfig(ctx)
+	if err != nil {
+		logger.Fatal("Error loading observability configuration: ", err)
+	}
+	observabilityConfigMap := observabilityConfig.GetConfigMap()
+	metrics.ConfigMapWatcher(ctx, component, SecretFetcher(ctx), logger)(&observabilityConfigMap)
+	profilingHandler.UpdateFromConfigMap(&observabilityConfigMap)
```
Comment (mattmoor) [inline injection/sharedmain/main.go:277]: For parity with logging, I'd probably move this into a function like this one: https://github.com/knative/pkg/blob/714b7630a836ef3952826849f31fe9bc16493218/injection/sharedmain/main.go#L311-L313

Comment (k4leung4) [reply injection/sharedmain/main.go]: done, added function for setting up observability

**Review by mattmoor — APPROVED** [review]
thanks!

/lgtm
/approve

**Pull request merged**